### ISSUE TITLE
Blob Storage State Http Conditional Header Not Met Exception

### ISSUE THREAD

**Issue by rosskyl** [Microsoft/botbuilder-dotnet#1292]
## Version
C# 4.0.8

## Describe the bug
Every once in a while we get an exception traced back to the middleware where we call next. It looks like we are getting a 412 response back from Azure Blob Storage.

## Expected behavior
No exception.

## Additional context
Exception:

```
Exception caught : System.AggregateException: One or more errors occurred. (The condition specified using HTTP conditional header(s) is not met.) ---> Microsoft.WindowsAzure.Storage.StorageException: The condition specified using HTTP conditional header(s) is not met.
 at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.ExecuteAsyncInternal[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Core\Executor\Executor.cs:line 316
 at Microsoft.WindowsAzure.Storage.Blob.CloudBlob.DownloadRangeToStreamAsync(Stream target, Nullable`1 offset, Nullable`1 length, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext, IProgress`1 progressHandler, CancellationToken cancellationToken) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Blob\CloudBlob.cs:line 438
 at Microsoft.WindowsAzure.Storage.Blob.BlobReadStream.DispatchReadAsync(Byte[] buffer, Int32 offset, Int32 count) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Blob\BlobReadStream.cs:line 137
 --- End of inner exception stack trace ---
 at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
 at Microsoft.WindowsAzure.Storage.Blob.BlobReadStream.<>c__DisplayClass4_0.<Read>b__0() in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Blob\BlobReadStream.cs:line 80
 at Microsoft.WindowsAzure.Storage.Core.Util.CommonUtility.RunWithoutSynchronizationContext[T](Func`1 actionToRun) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\Common\Core\Util\CommonUtility.cs:line 402
 at System.IO.StreamReader.ReadBuffer(Span`1 userBuffer, Boolean& readToUserBuffer)
 at System.IO.StreamReader.ReadSpan(Span`1 buffer)
 at System.IO.StreamReader.Read(Char[] buffer, Int32 index, Int32 count)
 at Newtonsoft.Json.JsonTextReader.ReadData(Boolean append, Int32 charsRequired)
 at Newtonsoft.Json.JsonTextReader.ParseValue()
 at Newtonsoft.Json.JsonReader.ReadForType(JsonContract contract, Boolean hasConverter)
 at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
 at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
 at Microsoft.Bot.Builder.Azure.AzureBlobStorage.<>c__DisplayClass7_0.<<ReadAsync>g__ReadIndividualKey|4>d.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder.Azure\AzureBlobStorage.cs:line 135
 --- End of stack trace from previous location where exception was thrown ---
 at Microsoft.Bot.Builder.Azure.AzureBlobStorage.ReadAsync(String[] keys, CancellationToken cancellationToken) in D:\a\1\s\libraries\Microsoft.Bot.Builder.Azure\AzureBlobStorage.cs:line 115
 at Microsoft.Bot.Builder.BotState.LoadAsync(ITurnContext turnContext, Boolean force, CancellationToken cancellationToken) in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotState.cs:line 67
 at Microsoft.Bot.Builder.BotState.BotStatePropertyAccessor`1.GetAsync(ITurnContext turnContext, Func`1 defaultValueFactory, CancellationToken cancellationToken) in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotState.cs:line 284
 at Microsoft.Bot.Builder.Dialogs.DialogSet.CreateContextAsync(ITurnContext turnContext, CancellationToken cancellationToken) in D:\a\1\s\libraries\Microsoft.Bot.Builder.Dialogs\DialogSet.cs:line 64
 at EdinaRealty.Emma.Bot.v4.EmmaBot.OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken) in F:\Development_New\EdinaRealty.Emma\EdinaRealty.Emma\EdinaRealty.Emma.Bot.v4\EmmaBot.cs:line 102
 at EdinaRealty.Emma.Bot.v4.Middleware.SpeakSettingMiddleware.OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken) in F:\Development_New\EdinaRealty.Emma\EdinaRealty.Emma\EdinaRealty.Emma.Bot.v4\Middleware\SpeakSettingMiddleware.cs:line 42
 at Microsoft.Bot.Builder.MiddlewareSet.ReceiveActivityWithStatusAsync(ITurnContext turnContext, BotCallbackHandler callback, CancellationToken cancellationToken) in D:\a\1\s\libraries\Microsoft.Bot.Builder\MiddlewareSet.cs:line 55
 at Microsoft.Bot.Builder.BotAdapter.RunPipelineAsync(ITurnContext turnContext, BotCallbackHandler callback, CancellationToken cancellationToken) in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotAdapter.cs:line 167
 ---> (Inner Exception #0) Microsoft.WindowsAzure.Storage.StorageException: The condition specified using HTTP conditional header(s) is not met.
 at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.ExecuteAsyncInternal[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Core\Executor\Executor.cs:line 316
 at Microsoft.WindowsAzure.Storage.Blob.CloudBlob.DownloadRangeToStreamAsync(Stream target, Nullable`1 offset, Nullable`1 length, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext, IProgress`1 progressHandler, CancellationToken cancellationToken) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Blob\CloudBlob.cs:line 438
 at Microsoft.WindowsAzure.Storage.Blob.BlobReadStream.DispatchReadAsync(Byte[] buffer, Int32 offset, Int32 count) in C:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\WindowsRuntime\Blob\BlobReadStream.cs:line 137
 Request Information
 RequestID:d0b2a7f0-801e-000e-26a2-ae3bd9000000
 RequestDate:Thu, 17 Jan 2019 14:22:07 GMT
 StatusMessage:The condition specified using HTTP conditional header(s) is not met.
 ErrorCode:ConditionNotMet
 ErrorMessage:The condition specified using HTTP conditional header(s) is not met. RequestId:d0b2a7f0-801e-000e-26a2-ae3bd9000000 Time:2019-01-17T20:22:07.1367715Z
 <---
```

[bug]

**Comment by tdurnford** [comment]
I would recommend upgrading your BotBuilder and BotFramework NuGet Packages to the newest release. There have been several releases since v4.0.8 - the most recent version being v4.2.0 which was released last month.

**Comment by johnataylor** [comment]
This behavior does appear to be consistent with what I'm seeing in the code, both for JavaScript and C#.

Specifically the Read implementations get the eTag and the write uses it if its there. So if there is some concurrency you should see exceptions.

However, I don't think this was the intended behavior, the intended behavior was "last one wins" perhaps @Stevenic can comment.

**Comment by johnataylor** [comment]
Looking more closely at the JavaScript in the higher layer BotState saveChange it sets the eTag to "*" and this is the protocol to talk to the storage plug ins. The C# doesn't do this and it was the intention of the design that it should. So this is a bug in C#.

Unfortunately the interface to the storage layer is a little convoluted expressed in terms of object which might be a IStoreItem but might not.

The simplest fix might be to default to "*" if its not an IStoreItem.

**Comment by johnataylor** [comment]
Drilling into this a little I don't think this has anything to do with the IStoreItem ETag == "*" behavior the AzureBlobStorage code builds in. Firstly that code seems to basically work, secondly the ConversationState doesn't use that anyhow (because it saves a dictionary NOT an IStoreItem) and third this exception occurs on the read path anyhow!

However, looking closely at this code and it appears to have a couple of optimizations both of which would cause the Blob Storage Client to be used from multiple threads - this is absolutely wrong. That client is not thread safe.

In particular the read creates a collection of Tasks and then waits on them. This is wrong, each task would need its own blob storage client.

And then overall the class attempts to cache a _container again this is wrong as this class is often used as a singleton and so again could potentially have multiple threads.

So the fix here is to remove these optimizations.

**Comment by drub0y** [comment]
### Root Cause

The `412` status during a blob read happens if the blob is modified while being downloaded which can happen if two threads/servers are processing a turn for the conversation at the same time with one reading and the other writing. As such, this is just an Azure Storage specific manifestation of a concurrency issue.

### A Simple Repro

[Here's an app that took about 20mins to whip together that succintly repros the issue](https://github.com/drub0y/BotBuilder.AzureBlobStorage.412Repro) using nothing but the `AzureBotStorage` APIs. The project references the released `4.2.2` version of the package. However, if you manually reference the newer implementation with the changes made by the PR **you will still observe the same issue**.

### AccessCondition not the problem

There was, AFAICT, nothing wrong with the `AccessCondition` logic. The "fix" provided by the PR simply flips the ternary condition around, resulting in no actual difference in behavior:

![issue1292](https://user-images.githubusercontent.com/284152/52811075-108e0700-3049-11e9-8189-77bd551ecc33.png)

### Regressions caused by the PR

1. The changes have reverted all the concurrent reading/writing around individual storage keys, which is detrimental to throughput because it serializes work that could otherwise be done in parallel.

2. The changes have introduced an additional HTTP request to Azure Storage _for every call_ to the `AzureBlobStorage::WriteAsync` API in the form of constantly calling `CreateIfNotExistsAsync` to ensure the container exists. So basically, every on turn that saves state you will be making a minimum of two calls instead of one while the state is being persisted. This not only costs developers runtime performance in terms of network latency, but will cost actual 💰 in terms calls to Azure Storage.

### Azure Storage SDK concurrency concerns

Despite the standard, boilerplate notes about thread safety on  the `CloudBlobClient` and related API docs, in practice they are known to be completely thread safe.

I have built and shipped some highly scalable, highly concurrent Azure Storage based solutions utilizing them in this way (for several years) and have _never_ had an issue with the SDK in this regard. You can peruse the source code and see for yourself that the implementation is thread safe. Also, if there _were_ issues at the SDK level with how `AzureBlobStorage` is using the APIs, I think those would have been surfaced when Southworks performed stress testing on the implementation or even now in my repro app.

And, yes, [I am fully aware I am contradicting myself from 6yrs ago](https://stackoverflow.com/a/9934418/185859), but the SDK has changed _drastically_ since then. 😆

### There is no "real" fix for this

There is no automatic, one size all fix for concurrency issues. From an SDK perspectice, a good solution for this kind of problem would be to catch the `412` storage exception and translate that to an abstract exception that is married to the `IStorage` abstraction layer (e.g. an `StorageConcurrencyException`) with the  `AzureBlobStorage` doing that translation to when it encounters the `412`. Likewise there will be concurrency issues with any other storage layer, such as `CosmosDbStorage` which would watch for its specific representation of this error and translate it to the same abstract exception.

This exception would next bubble up from the storage abstraction through the bot state abstraction in `BotState::LoadAsync` where, technically, it should be translated to another exception specific to that level of abstraction (e.g. an `BotStateConcurrencyException`). Next, that would bubble up through the `IStatePropertyAccessor<T>::GetAsync` and, technically, another abstraction would be good here too since the whole point of this interface is to decouple from the `BotState` implementation. Finally it would reach the bot/app layer where it could be caught and acted upon however the developer sees fit such as maybe retrying the call to `GetAsync`.

### Final recommendations

* I would _highly_ recommend reverting the changes made in this PR based on the aforementioned regressions it has introduced when absolutely nothing was/is provably wrong with the existing implementation.
* This issue regarding the `412` status should be reopened until it is properly addressed.

**Comment by johnataylor** [comment]
in 4.3 the blob storage provider is much improved and this issue (on the read) is addressed with retry.
